modules/checkout/src/Controller/CheckoutController.php

```diff
@@ -119,4 +142,98 @@ public function checkAccess(RouteMatchInterface $route_match, AccountInterface $
     return $access;
   }
 
+  /**
+   * Checks access to a particular checkout page.
+   *
+   * @param Drupal\commerce_order\Entity\OrderInterface $order
+   *   The fully loaded order object represented on the checkout form.
+   *
+   * @return bool
+   *   TRUE or FALSE indicating access.
+   */
+  protected function checkoutPageAccess(OrderInterface $order) {
+    $checkout_flow = $this->checkoutOrderManager->getCheckoutFlow($order);
+    $requested_step = $checkout_flow->getPlugin()->getStepId();
+    $visible_steps = $checkout_flow->getPlugin()->getVisibleSteps();
+    $visible_step_ids = array_keys($visible_steps);
+    $first_step = reset($visible_step_ids);
+    $order_step = $this->getOrderCheckoutStep($order);
+
+    // If the order is not in checkout, return FALSE for any page but the
+    // completion page.
+    if ($order_step === 'complete' && $requested_step !== 'complete') {
```

Comment:
We should improve this to take into account the order workflow.
- If the order has been placed, allow the "complete" step (and only the "complete" step).
- If the order is still a draft, allow the current step and the previous ones.
Otherwise 403. That covers both cancelled and refunded.

Thoughts:
- I guess it's safe to assume that every checkout flow will have a "complete" step (but how do we enforce that?)
- We need to implement the "payment" use case somehow, where once you reach the payment step you can't go back to the previous steps, and the order in checkout is no longer considered a cart.